### PR TITLE
[add] config

### PR DESCRIPTION
**Pull request by chunibyo-wly**
But I don't know why we need to freeze a config object?
It's looked weired forcing use merge to update configuration?

### CODE REVIEW

**Comment by odashi** [comment]
Ref: #66 

@chunibyo-wly Thanks!

Just a comment about "why we need to freeze": it is basically because immutability keeps ease of maintainability and reproducibility.
Always generating a fresh struct keeps the entire library unchaged, i.e., we can guarantee `latexify.function` has no side effect and we could get exactly the same result every time we called `latexify.function` even it is invoked from multiple threads. Immutability is also applied in the entire process of this library, especially tree transformation chains in `get_latex`: all transformers always return a new AST (with references to possibly old subtrees for efficiency).

We may add some functionality to override the "global" config in the future. This kind of functionality should be explicitly mentioned that "invoking this method causes side effect and not thread safe."

File: src/latexify/config.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from __future__ import annotations
```
Comment (odashi) [inline src/latexify/config.py:1]:
```suggestion
"""Definition of the Config class."""

from __future__ import annotations

```

File: src/latexify/config.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import dataclasses
+
+from typing import Any
+
+
+@dataclasses.dataclass(frozen=True)
+class Config:
+    identifiers: dict[str, str] | None
+    expanded_functions: set[str] | None
```
Comment (odashi) [inline src/latexify/config.py:10]: This is not an available config at this point.
```suggestion
```

File: src/latexify/config.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+from __future__ import annotations
+import dataclasses
+
+from typing import Any
+
+
+@dataclasses.dataclass(frozen=True)
+class Config:
```
Comment (odashi) [inline src/latexify/config.py:8]:
```suggestion
class Config:
   """Configurations to control the behavior of latexify."""

```

File: src/latexify/config.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    identifiers: dict[str, str] | None
+    expanded_functions: set[str] | None
+    reduce_assignments: bool
+    use_math_symbols: bool
+    use_raw_function_name: bool
+    use_signature: bool
+    use_set_symbols: bool
+
+    def merge(self, *, config: Config | None = None, **kwargs) -> Config:
```
Comment (odashi) [inline src/latexify/config.py:17]: Add docstring for all public functions. It is required to follow the Google's style: https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: src/latexify/config.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            return arg
+
+        return Config(**{f.name: merge_field(f.name) for f in dataclasses.fields(self)})
+
+    def expand_function(self, fun: str | list[str]) -> None:
+        if type(fun) == str:
+            self.expanded_functions.add(fun)
+        elif type(fun) == list:
+            self.expanded_functions.update(fun)
+
```
Comment (odashi) [inline src/latexify/config.py:35]: remove this method.
```suggestion
```

File: src/latexify/config.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    @staticmethod
+    def defaults() -> Config:
+        return Config(
+            identifiers=None,
+            expanded_functions=set(),
```
Comment (odashi) [inline src/latexify/config.py:40]:
```suggestion
```

File: src/latexify/frontend.py [hunk 7]
```diff
@@ -10,18 +10,10 @@
...
-    reduce_assignments: bool = False,
-    use_math_symbols: bool = False,
-    use_raw_function_name: bool = False,
-    use_signature: bool = True,
-    use_set_symbols: bool = False,
-) -> str:
+def get_latex(fn: Callable[..., Any], *, config: Config | None = None, **kwargs) -> str:
     """Obtains LaTeX description from the function's source.
 
     Args:
```
Comment (odashi) [inline src/latexify/frontend.py:19]: Add appropriate "Args" section.

I also think that this section should have only descriptions about `config` and `**kwargs`. Descriptions of specific configs can be written in the `Config` class instead.

File: src/latexify/config.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            use_math_symbols: Whether to convert identifiers with a math symbol surface
+                (e.g., "alpha") to the LaTeX symbol (e.g., "\\alpha").
+            use_raw_function_name: Whether to keep underscores "_" in the function name,
+                or convert it to subscript.
+            use_signature: Whether to add the function signature before the expression
+                or not.
+            use_set_symbols: Whether to use set symbols or not.
+
+        Returns:
+            a new thread safe Config object
```
Comment (odashi) [inline src/latexify/config.py:43]: I think this object itself does not guarantee the thread safety.
```suggestion
            A new Config object
```

File: src/latexify/config.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+                if config is not None:
+                    arg = getattr(config, name)
+                else:
+                    arg = getattr(self, name)
+            return arg
+
+        return Config(**{f.name: merge_field(f.name) for f in dataclasses.fields(self)})
+
+    @staticmethod
+    def defaults() -> Config:
```
Comment (odashi) [inline src/latexify/config.py:59]:
```suggestion
    def defaults() -> Config:
        """Generates a Config with default values.
        
        Returns:
            A new Config with default values
        """
```

File: src/latexify/frontend.py [hunk 10]
```diff
@@ -10,58 +10,41 @@
 from latexify import exceptions
 from latexify import parser
 from latexify import transformers
+from latexify.config import Config
```
Comment (odashi) [inline src/latexify/frontend.py:13]: Google's style guide does not allow direct imports.

https://google.github.io/styleguide/pyguide.html#22-imports

```suggestion
from latexify import config
```

File: src/latexify/frontend.py [hunk 11]
```diff
@@ -10,58 +10,41 @@
...
-        use_math_symbols: Whether to convert identifiers with a math symbol surface
-            (e.g., "alpha") to the LaTeX symbol (e.g., "\\alpha").
-        use_raw_function_name: Whether to keep underscores "_" in the function name,
-            or convert it to subscript.
-        use_signature: Whether to add the function signature before the expression or
-            not.
-        use_set_symbols: Whether to use set symbols or not.
+        config: use defined Config object, if it is None, it will be automatic assigned
+            with default value.
+        kwargs: dict of Config field values that could be defined individually by uses.
```
Comment (odashi) [inline src/latexify/frontend.py:23]: `kwargs` looks there is the exact parameter `kwargs`.

```suggestion
        **kwargs: dict of Config field values that could be defined individually by uses.
```

File: src/latexify/config.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+from __future__ import annotations
+
+import dataclasses
+
+from typing import Any
+
+
+@dataclasses.dataclass(frozen=True)
+class Config:
+    """Configurations to control the behavior of latexify."""
```
Comment (odashi) [inline src/latexify/config.py:12]: Could you put descriptions of members here?

```suggestion
    """Configurations to control the behavior of latexify.
    
    Attributes:
        foo: ...
        bar: ...
    """
```

File: src/latexify/config.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+                ^[A-Za-z_][A-Za-z0-9_]*$
+            reduce_assignments: If True, assignment statements are used to synthesize
+                the final expression.
+            use_math_symbols: Whether to convert identifiers with a math symbol surface
+                (e.g., "alpha") to the LaTeX symbol (e.g., "\\alpha").
+            use_raw_function_name: Whether to keep underscores "_" in the function name,
+                or convert it to subscript.
+            use_signature: Whether to add the function signature before the expression
+                or not.
+            use_set_symbols: Whether to use set symbols or not.
```
Comment (odashi) [inline src/latexify/config.py:40]:
```suggestion
            **kwargs: Members to modify. This value precedes both self and config.
```

**Review by odashi — COMMENTED** [review]
Almost fine, and there are only trivial comments remaining.

**Comment by odashi** [comment]
By the way, there are several pull requests that may affect config so there would be some collisions when merging. I guess it is better to merge this pull request first, and request others to merge `main` after merging this one.

**Comment by odashi** [comment]
@chunibyo-wly Hi, I like this change, and would like to merge it soon! If you could resolve the comments above, I really appreciate it.

**Comment by chunibyo-wly** [comment]
sorry, I forgot to check my email

File: src/latexify/frontend.py (hunk 10 above)
Comment (chunibyo-wly) [reply src/latexify/frontend.py:13]: sorry, I don't know how to modify this one.
Could you help me?

Comment (ZibingZhang) [reply src/latexify/frontend.py:13]: you'll have to change the import, and change any uses of `Config` to `config.Config`

Comment (chunibyo-wly) [reply src/latexify/frontend.py:13]:
```python
def get_latex(
    fn: Callable[..., Any], *, _config: config.Config | None = None, **kwargs
) -> str:
```

Thanks for your reply. Is it looks like this one?

Comment (ZibingZhang) [reply src/latexify/frontend.py:13]: Ah yeah that makes it a bit awkward for the keyword argument doesn't it.

Maybe change the import to `from latexify import config as cfg`?

That way we can keep the name of the keyword argument as `..., config: cfg.Config | None = None, ...`.

Comment (chunibyo-wly) [reply src/latexify/frontend.py:13]: Thanks for your advice, you were right

**Review by ZibingZhang — CHANGES_REQUESTED** [review]
Since https://github.com/google/latexify_py/pull/125 has been merged in, there's a new argument `expand_functions` in `get_latex`.

Could you update the `Config` class to include that as well?

**Comment by odashi** [comment]
@ZibingZhang Yeah, but it could be processed after merging this PR.

**Review by odashi — APPROVED** [review]
I'd approve this. maybe we need a follow-up PR to introduce a new flag (or directly applying it here)

**Comment by odashi** [comment]
I applied a monkey patch to remain `expand_functions` as is to merge this PR. After fixing the linter error I will merge it.

**Comment by odashi** [comment]
@chunibyo-wly Thanks for your work! I think this is very important change for this repository.

**Pull request merged**